
*  DSC 530
*  Week 3 & 4 Assignment
*  Saebastion Cole

# Hands-on data analysis with Pandas

# Chapter 1, Exercise 4

Run the code in the first cell of the exercises.ipynb notebook. It will give you a list of 100 values to work with for the rest of the exercises in this chapter. Be sure to treat these values as a sample of the population.


In [1022]:
import random

random.seed(0)
salaries = [round(random.random()*1000000, -3) for _ in range(100)]
# Currently, this code generated a list of random values as salaries for this exercise.

## Chapter 1, Exercise 5

Using the data from exercise 4, calculate the following statistics without importing anything from the [statistics module](https://docs.python.org/3/library/statistics.html) in the Python standard library, and then confirm your results match up to those that are obtained when using the statistics module (where possible):


Tip from instructor: You can use [Pandas](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html) or [Numpy](https://numpy.org/doc/stable/reference/routines.statistics.html) for the initial calculation and then check against other Python library/package such as [statistics module](https://docs.python.org/3/library/statistics.html), or [SciPy Statistics](https://docs.scipy.org/doc/scipy/reference/stats.html). Part a of Exercise 5 is completed to illustrate the process for you, you can decide how to go about the other problems.

Exercise 5(a) Mean - with exmaples

In [1023]:
# Run this code to make alias to shorted the typed name
# You can do this as the 1st thing in any notebook or later, but you only need to do this once
import numpy as np
import pandas as pd

In [1024]:


# Using numpy.mean
np.mean(salaries)

np.float64(585690.0)

In [1025]:
# Using statistics module
import statistics # no need to re-run this after import
# You have to import this standard library before using its mean function
statistics.mean(salaries)

585690.0

In [1026]:
# Using Pandas
# If you choose to use Pandas functions, you have to convert the list to a Series, otherwise it will gives an error
salaries_series = pd.Series(salaries)
# For Pandas function, when you placed a dot after a variable or other functions, it is referred as a "method chaining" appraoch.
salaries_series.mean()

np.float64(585690.0)

In [1027]:
# Using SciPy Statistics
from scipy import stats
stats.tmean(salaries)

np.float64(585690.0)

Exercise 5(b) Median

In [1028]:
# Your 1st appraoch

np.median(salaries)


np.float64(589000.0)

In [1029]:
# Using Pandas
salaries_series.median()

np.float64(589000.0)

In [1030]:
# Using statistics module
statistics.median(salaries)

589000.0

Exercise 5(c) Mode

Hint from textbook: Check out the [Counter class](https://docs.python.org/3/library/collections.html#collections.Counter) in the collections module of the standard library

In [1031]:
# Your 1st approach
# use np.unique to get the unique values and their counts
vals, counts = np.unique(salaries, return_counts=True)
# Return the value with the highest count and the counts 
mode = vals[counts.max()], counts.max()
mode

(np.float64(82000.0), np.int64(3))

In [1032]:
# Using Pandas

# This way only shows the mode value
print(salaries_series.mode())

# To show count as well you need value_counts()
# I use head(1) to show only the first row
print(salaries_series.value_counts().head(1))

0    477000.0
dtype: float64
477000.0    3
Name: count, dtype: int64


In [1033]:


from collections import Counter

# Using counter to find the most common and mode
counter_mode = Counter(salaries).most_common(1)


counter_mode[0]

(477000.0, 3)

In [1034]:
# Using statistics module
# Statistics only will show the value not the count so I will use the counter
stats_mode   = statistics.mode(salaries), Counter(salaries).most_common(1)[0][1]
stats_mode


(477000.0, 3)

Exercise 5(d) Sample Variance

In [1035]:
# Your 1st approach
np.var(salaries, ddof=1)

np.float64(70664054444.44444)

In [1036]:
# Using Pandas
# I have to use a series instead of a list
salaries_series.var()

np.float64(70664054444.44444)

In [1037]:
# Using statistics module
statistics.variance(salaries)

70664054444.44444

Exercise 5(e) Sample Standard Deviation

In [1038]:
# Your 1st approach
np.std(salaries, ddof=1)

np.float64(265827.11382484)

In [1039]:
# Using Pandas
salaries_series.std()

np.float64(265827.11382484)

In [1040]:
# Using statistics module
statistics.stdev(salaries)

265827.11382484

## Chapter 1, Exercise 6

Using the data from exercise 4, calculate the following statistics using the functions in the statistics module where appropriate:

Exercise 6(a) Range

In [1041]:
# Hint:there is no function called "Range". However, what does it means to see a range of values?
# Your solution

# To get the range we subtract the minimum value from the maximum value
salaries_range = max(salaries) - min(salaries)
salaries_range

995000.0

Exercise 6(b) Coefficient of Variation

In [1042]:
# Hint: Coefficient of Variation (CV) is defined as the ratio of standard deviation to the mean in Hands-on Data Analysis with Pandas.
# Your solution


# Divide the standard deviation by the mean
salaries_cv = salaries_series.std() / salaries_series.mean()
salaries_cv

np.float64(0.45386998894439035)

Exercise 6(c) Interquartile Range

In [1043]:
# Hint: Interquartile range (IQR) is presented as the difference of the 75% and 25% quantile in Chapter 1 of Practical Statistics for Data Scientisits Textbook
# OR you can find the values in each quartile and then substract Quantile 3 with Quantile 1 as in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution

# IQR is the quartile 3 minus the quartile 1
iqr = salaries_series.quantile(0.75) - salaries_series.quantile(0.25)
iqr

np.float64(413250.0)

Exercise 6(d) Quartile coefficient of dispersion

In [1044]:
# Once you determined the 1st and 3rd quantile values from 6(c), you can use it for this calculation
# Refer to the Quartile coefficient of dispersion formula in Chapter 1 of Hands-on Data Analysis with Pandas
# Your solution

# Use the iqr and divide it by the sum of the 1st and 3rd quantile values
iqr_coefficient_of_dispersion = iqr / (salaries_series.quantile(0.75) + salaries_series.quantile(0.25))
iqr_coefficient_of_dispersion

np.float64(0.338660110633067)

## Chapter 1, Exercise 7

Scale the data created in exercise 4 using the following strategies:

Exercise 7(a) Min-max Scaling (Normalizing)

In [1045]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the scikit-learn function "MinMaxScaler", which requires transforming the data in 2D array before applying the scaler. This is a common approach in machine learning.
# Your solution

# So for this using pandas we need to convert the data to a series then resize it to a 0 - 1 range to normalize the data
normalized_salaries = (salaries_series - salaries_series.min()) / (salaries_series.max() - salaries_series.min())
normalized_salaries

0     0.847236
1     0.760804
2     0.422111
3     0.259296
4     0.512563
        ...   
95    0.920603
96    0.795980
97    0.081407
98    0.615075
99    0.487437
Length: 100, dtype: float64

Exercise 7(b) Standardizing

In [1046]:
# Refer to the Scaling data section of your reading in Chapter 1 of Hands-on Data Analysis with Pandas
# You can calculate this value by following the formula in the textbook or see the example in the solution
# You can also test out the z-score method in Scipy Stats (2D array is needed)
# Your solution

# Very similar to the normalization question but instead of basing the data on the mena we base it on the standard deviation. This is called standardization
standardized_salaries = (salaries_series - salaries_series.mean()) / salaries_series.std()
standardized_salaries

0     0.971722
1     0.648203
2    -0.619538
3    -1.228957
4    -0.280972
        ...   
95    1.246336
96    0.779868
97   -1.894803
98    0.102736
99   -0.375018
Length: 100, dtype: float64

## Chapter 1, Exercise 8

Using the scaled data from exercise 7, calculate the following:

Note. There is some preparation work to create the normalized (scaled) vs. standardized dataset before calculating covariance.

In [1047]:
# If you did not setup a scaled data or standardized data with z-score approach, you can use the following codes.
# Scaled data aka normalized data
from sklearn.preprocessing import MinMaxScaler
# Make an array of salaries
salaries_array = np.array(salaries)
# Reshape the data to 2D (one feature)
salaries_2d = salaries_array.reshape(-1, 1)
# Scaled the data
scaler = MinMaxScaler()
scaler.fit(salaries_2d)
# Optionally, transform the data. scaled_salaries in the scaled data for use in next step
scaled_salaries = scaler.transform(salaries_2d)
# Preview the first 10 values. Noticed the double [] that shows the 2D shape
scaled_salaries[:10]
# This will print the shape if you would like to check
# scaled_salaries.shape

array([[0.84723618],
       [0.76080402],
       [0.42211055],
       [0.25929648],
       [0.51256281],
       [0.40603015],
       [0.78693467],
       [0.30351759],
       [0.47839196],
       [0.58492462]])

In [1048]:
# Standardized version of the salary data (z-score)
from scipy import stats
# Assign a new variable `standardized_salary` to store the standarized data
standardized_salary = stats.zscore(salaries_array)
# Preview the first 10 values. This is a 1 dimension object
standardized_salary[:10]
# This will print the shape if you would like to check
# standardized_salary.shape

array([ 0.97661715,  0.65146878, -0.62265912, -1.23514791, -0.28238758,
       -0.68315184,  0.74976945, -1.06879293, -0.41093461, -0.01017034])

Exercise 8(a) The covariance between the standardized and normalized data

In [1049]:
# Covariance can be calculated with numpy
# When the (normalized) scaled_salaries varaible was created, it was created as a 2D array.
# We will have to 'flatten' it to match the shape of the standardized_salary before calculation.
scaled_salaries = np.array(scaled_salaries).flatten()

# Now calculate the covariance
cov_calculated = np.cov(scaled_salaries, standardized_salary)
print(cov_calculated)

[[0.07137603 0.26850885]
 [0.26850885 1.01010101]]


Note: How to interpret this output? The covariance function produce a 2X2 covariance matrix, the off-diagonal elements present the covariance of the 2 sets of values. Therefore 0.2685 is the output for intrepretation. Since the covariance estimate is positive, it indicates a positive relationship (moving in the same direction), and we don't know how "strong" this relationship and therefore correlation will help us understand this better.

In [1050]:
# Example 2 for covariance calculation
# Covariance can be calculated with statistics module
statistics.covariance(scaled_salaries, standardized_salary)

np.float64(0.2685088459449036)

### My Approach

In [1051]:
cov = np.cov(normalized_salaries, standardized_salary)

cov

array([[0.07137603, 0.26850885],
       [0.26850885, 1.01010101]])

Exercise 8(b) The Pearson correlation coefficient between the standardized and normalized data (this is actually 1, but due to rounding along the way, the result will be slightly less)

In [1052]:
# Correlation can be calculated with statistics module
statistics.correlation(scaled_salaries, standardized_salary)
# The result is 1 because we are checking relationship of the same data in different format.

np.float64(1.0000000000000002)

Note. Getting a correlation of 1 is often very rare, but in this case, we are basically using the same raw data and simply normalized it (force values to 0,1 range) or standardized it (force the data to centered around zero and then deviate from zero) to put it on a common scale. It will not change the strength or direction of the relationship between 2 variables.

### My Approach

In [1053]:
normalized_salaries.corr(standardized_salaries)

np.float64(1.0)

# Chapter 2

In [1054]:
# The default code to run if you setup your book_env correctly to import the dataset
df = pd.read_csv('./data/parsed.csv')

df.head(5)
# If you encountered issues due to directory setup, you can provide the full file path also
# df = pd.read_csv('cdrive../../ch_02/data/parsed.csv')

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
0,NaN,NaN,37389218,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.008693,NaN,85.0,",ci37389218,",1.35,ml,...,automatic,1539475168010,"M 1.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475395144,https://earthquake.usgs.gov/earthquakes/eventp...,California
1,NaN,NaN,37389202,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.020030,NaN,79.0,",ci37389202,",1.29,ml,...,automatic,1539475129610,"M 1.3 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475253925,https://earthquake.usgs.gov/earthquakes/eventp...,California
2,NaN,4.4,37389194,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.021370,28.0,21.0,",ci37389194,",3.42,ml,...,automatic,1539475062610,"M 3.4 - 8km NE of Aguanga, CA",0,earthquake,",dyfi,focal-mechanism,geoserve,nearby-cities,o...",-480.0,1539536756176,https://earthquake.usgs.gov/earthquakes/eventp...,California
3,NaN,NaN,37389186,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.026180,NaN,39.0,",ci37389186,",0.44,ml,...,automatic,1539474978070,"M 0.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475196167,https://earthquake.usgs.gov/earthquakes/eventp...,California
4,NaN,NaN,73096941,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.077990,NaN,192.0,",nc73096941,",2.16,md,...,automatic,1539474716050,"M 2.2 - 10km NW of Avenal, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,scit...",-480.0,1539477547926,https://earthquake.usgs.gov/earthquakes/eventp...,California


## Chapter 2, Exercise 1

Find the 95th percentile of earthquake magnitude in Japan using the mb magnitude type.

In [1055]:
# Get the data filtered
jp_mb_data = df[(df.parsed_place == 'Japan') & (df.magType == 'mb')]
jp_mb_data.head(5)

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
67,NaN,NaN,1000hbqa,https://earthquake.usgs.gov/fdsnws/event/1/que...,1.415,NaN,72.0,",us1000hbqa,",4.6,mb,...,reviewed,1539448501800,"M 4.6 - 160km NNW of Nago, Japan",0,earthquake,",geoserve,origin,phase-data,",480.0,1539449501040,https://earthquake.usgs.gov/earthquakes/eventp...,Japan
713,NaN,NaN,1000hah8,https://earthquake.usgs.gov/fdsnws/event/1/que...,1.141,NaN,82.0,",us1000hah8,",4.7,mb,...,reviewed,1539238726290,"M 4.7 - 139km WSW of Naze, Japan",0,earthquake,",geoserve,origin,phase-data,",540.0,1539240344040,https://earthquake.usgs.gov/earthquakes/eventp...,Japan
1124,NaN,NaN,1000h9la,https://earthquake.usgs.gov/fdsnws/event/1/que...,1.737,NaN,135.0,",us1000h9la,",4.6,mb,...,reviewed,1539115120470,"M 4.6 - 53km ESE of Kamaishi, Japan",0,earthquake,",geoserve,origin,phase-data,",540.0,1539119067040,https://earthquake.usgs.gov/earthquakes/eventp...,Japan
1309,NaN,NaN,1000h952,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.897,NaN,136.0,",us1000h952,",4.5,mb,...,reviewed,1539058606580,"M 4.5 - 80km ENE of Misawa, Japan",0,earthquake,",geoserve,origin,phase-data,",540.0,1539060436040,https://earthquake.usgs.gov/earthquakes/eventp...,Japan
1398,NaN,NaN,1000h8zx,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.555,NaN,124.0,",us1000h8zx,",4.4,mb,...,reviewed,1539034912990,"M 4.4 - 65km NNE of Hachijo-jima, Japan",0,earthquake,",geoserve,origin,phase-data,",540.0,1539062940040,https://earthquake.usgs.gov/earthquakes/eventp...,Japan


In [1056]:
# Using numpy

np.percentile(jp_mb_data['mag'], 95)

np.float64(4.9)

In [1057]:
# Using Pandas
jp_mb_data['mag'].quantile(0.95)

np.float64(4.9)

## Chapter 2, Exercise 2

Find the percentage of earthquakes in Indonesia that were coupled with tsunamis.

In [1058]:
# Filter the data for indonesia
ind_tot_data = df[(df.parsed_place == 'Indonesia')]

ind_tot_data.head(5)


,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
9,NaN,NaN,1000hbtn,https://earthquake.usgs.gov/fdsnws/event/1/que...,3.191,NaN,37.0,",us1000hbtn,",4.7,mb,...,reviewed,1539472814760,"M 4.7 - 219km SSE of Saparua, Indonesia",0,earthquake,",geoserve,origin,phase-data,",540.0,1539473712040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
13,NaN,NaN,1000hbti,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.193,NaN,157.0,",us1000hbti,",4.5,mb,...,reviewed,1539470898340,"M 4.5 - 120km SSW of Banda Aceh, Indonesia",0,earthquake,",geoserve,origin,phase-data,",360.0,1539493516040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
180,green,NaN,1000hbhw,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.077,NaN,23.0,",us1000hbhw,",5.2,mww,...,reviewed,1539405255580,"M 5.2 - 25km E of Bitung, Indonesia",0,earthquake,",geoserve,losspager,origin,phase-data,shakemap,",480.0,1539412565560,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
421,NaN,NaN,1000hayw,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.812,NaN,80.0,",us1000hayw,",4.7,mb,...,reviewed,1539331098920,"M 4.7 - 38km SSW of Nggongi Satu, Indonesia",0,earthquake,",geoserve,origin,phase-data,",480.0,1539332252040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
660,NaN,NaN,1000haks,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.612,NaN,90.0,",us1000haks,",4.4,mb,...,reviewed,1539258833830,"M 4.4 - 51km WSW of Kasiguncu, Indonesia",0,earthquake,",geoserve,origin,phase-data,",480.0,1539260052040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia


In [1059]:
# Filter the data for tsunami in Indonesia
ind_tsu_data = ind_tot_data[(ind_tot_data.tsunami == 1)]


ind_tsu_data.head(5)

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
1406,NaN,NaN,1000h8zg,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.132,NaN,79.0,",us1000h8zg,",5.1,mb,...,reviewed,1539033346530,"M 5.1 - 13km E of Palu, Indonesia",1,earthquake,",geoserve,origin,phase-data,",480.0,1539037997040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
1698,NaN,4.3,1000h8gb,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.637,17.0,63.0,",us1000h8gb,",5.1,mww,...,reviewed,1538935041200,"M 5.1 - 14km ENE of Sambelia, Indonesia",1,earthquake,",dyfi,geoserve,origin,phase-data,",480.0,1539213200944,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
3112,NaN,NaN,1000h68e,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.745,NaN,80.0,",us1000h68e,",5.3,mb,...,reviewed,1538570285120,"M 5.3 - 29km SSW of Nggongi, Indonesia",1,earthquake,",geoserve,moment-tensor,origin,phase-data,",480.0,1538688306040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
3150,NaN,NaN,1000h66t,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.710,NaN,79.0,",us1000h66t,",5.0,mww,...,reviewed,1538560686080,"M 5.0 - 38km S of Nggongi Satu, Indonesia",1,earthquake,",geoserve,origin,phase-data,",480.0,1538561863040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia
3605,NaN,NaN,1000h5hb,https://earthquake.usgs.gov/fdsnws/event/1/que...,2.596,NaN,90.0,",us1000h5hb,",5.0,mb,...,reviewed,1538456366290,"M 5.0 - 50km WSW of Kasiguncu, Indonesia",1,earthquake,",geoserve,origin,phase-data,",480.0,1538457427040,https://earthquake.usgs.gov/earthquakes/eventp...,Indonesia


In [1060]:
# Compare the counts to get a percentage
percentage = (len(ind_tsu_data) / len(ind_tot_data)) * 100
percentage

23.12925170068027

## Chapter 2, Exercise 3

Calculate summary statistics for earthquakes in Nevada.

In [1061]:
# Filter the data for Nevada

nev_tot_data = df[(df.parsed_place == 'Nevada')]
nev_tot_data.head(5)

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,status,time,title,tsunami,type,types,tz,updated,url,parsed_place
32,NaN,NaN,00660878,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.091,NaN,155.59,",nn00660878,",1.3,ml,...,reviewed,1539461038285,"M 1.3 - 13km E of Hawthorne, Nevada",0,earthquake,",geoserve,origin,phase-data,",-480.0,1539482691389,https://earthquake.usgs.gov/earthquakes/eventp...,Nevada
39,NaN,NaN,00660886,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.095,NaN,155.25,",nn00660886,",-0.1,ml,...,reviewed,1539458844506,"M -0.1 - 6km NW of Lemmon Valley, Nevada",0,earthquake,",geoserve,origin,phase-data,",-480.0,1539482703428,https://earthquake.usgs.gov/earthquakes/eventp...,Nevada
41,NaN,NaN,00660885,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.097,NaN,156.31,",nn00660885,",0.2,ml,...,reviewed,1539458665088,"M 0.2 - 6km NW of Lemmon Valley, Nevada",0,earthquake,",geoserve,origin,phase-data,",-480.0,1539482702050,https://earthquake.usgs.gov/earthquakes/eventp...,Nevada
48,NaN,NaN,00660867,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.326,NaN,111.14,",nn00660867,",0.9,ml,...,reviewed,1539455306730,"M 0.9 - 42km WSW of Tonopah, Nevada",0,earthquake,",geoserve,origin,phase-data,",-480.0,1539482685124,https://earthquake.usgs.gov/earthquakes/eventp...,Nevada
49,NaN,NaN,00660884,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.097,NaN,155.98,",nn00660884,",-0.1,ml,...,reviewed,1539455017464,"M -0.1 - 6km NW of Lemmon Valley, Nevada",0,earthquake,",geoserve,origin,phase-data,",-480.0,1539482700579,https://earthquake.usgs.gov/earthquakes/eventp...,Nevada


In [1062]:
nev_tot_data.describe()

,cdi,dmin,felt,gap,mag,mmi,nst,rms,sig,time,tsunami,tz,updated
count,15.000000,681.000000,15.000000,681.000000,681.000000,1.00,681.000000,681.000000,681.000000,6.810000e+02,681.0,681.0,6.810000e+02
mean,2.440000,0.166199,2.400000,153.668120,0.500073,2.84,12.618209,0.151986,10.970631,1.538314e+12,0.0,-480.0,1.538402e+12
std,0.501142,0.166228,4.626013,68.735302,0.696710,NaN,9.866963,0.084662,19.607150,5.965637e+08,0.0,0.0,6.010951e+08
min,2.000000,0.001000,1.000000,29.140000,-0.500000,2.84,3.000000,0.000500,0.000000,1.537247e+12,0.0,-480.0,1.537307e+12
25%,2.000000,0.053000,1.000000,97.380000,-0.100000,2.84,6.000000,0.106900,0.000000,1.537854e+12,0.0,-480.0,1.537928e+12
50%,2.200000,0.112000,1.000000,149.140000,0.400000,2.84,10.000000,0.146300,2.000000,1.538280e+12,0.0,-480.0,1.538428e+12
75%,2.900000,0.233000,1.000000,199.720000,0.900000,2.84,16.000000,0.187100,12.000000,1.538821e+12,0.0,-480.0,1.538878e+12
max,3.300000,1.414000,19.000000,355.910000,2.900000,2.84,61.000000,0.863400,129.000000,1.539461e+12,0.0,-480.0,1.539483e+12


## Chapter 2, Exercise 4

Add a column indicating whether the earthquake happened in a country or US state that is on the Ring of Fire. Use Alaska, Antarctica (look for Antarctic), Bolivia, California, Canada, Chile, Costa Rica, Ecuador, Fiji, Guatemala, Indonesia, Japan, Kermadec Islands, Mexico (be careful not to select New Mexico), New Zealand, Peru, Philippines, Russia, Taiwan, Tonga, and Washington.

In [1063]:
ring_of_fire_loc = ['Alaska', 'Antarctic', 'Bolivia', 'California', 'Canada', 'Chile', 'Costa Rica',
                    'Ecuador', 'Fiji', 'Guatemala', 'Indonesia', 'Japan', 'Kermadec Islands', 'Mexico', 'New Zealand',
                    'Peru', 'Philippines', 'Russia', 'Taiwan', 'Tonga', 'Washington']

df['ring_of_fire'] = df.parsed_place.isin(ring_of_fire_loc).astype(int)
df.head(5)

,alert,cdi,code,detail,dmin,felt,gap,ids,mag,magType,...,time,title,tsunami,type,types,tz,updated,url,parsed_place,ring_of_fire
0,NaN,NaN,37389218,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.008693,NaN,85.0,",ci37389218,",1.35,ml,...,1539475168010,"M 1.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475395144,https://earthquake.usgs.gov/earthquakes/eventp...,California,1
1,NaN,NaN,37389202,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.020030,NaN,79.0,",ci37389202,",1.29,ml,...,1539475129610,"M 1.3 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475253925,https://earthquake.usgs.gov/earthquakes/eventp...,California,1
2,NaN,4.4,37389194,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.021370,28.0,21.0,",ci37389194,",3.42,ml,...,1539475062610,"M 3.4 - 8km NE of Aguanga, CA",0,earthquake,",dyfi,focal-mechanism,geoserve,nearby-cities,o...",-480.0,1539536756176,https://earthquake.usgs.gov/earthquakes/eventp...,California,1
3,NaN,NaN,37389186,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.026180,NaN,39.0,",ci37389186,",0.44,ml,...,1539474978070,"M 0.4 - 9km NE of Aguanga, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,",-480.0,1539475196167,https://earthquake.usgs.gov/earthquakes/eventp...,California,1
4,NaN,NaN,73096941,https://earthquake.usgs.gov/fdsnws/event/1/que...,0.077990,NaN,192.0,",nc73096941,",2.16,md,...,1539474716050,"M 2.2 - 10km NW of Avenal, CA",0,earthquake,",geoserve,nearby-cities,origin,phase-data,scit...",-480.0,1539477547926,https://earthquake.usgs.gov/earthquakes/eventp...,California,1


## Chapter 2, Exercise 5

Calculate the number of earthquakes in the Ring of Fire locations and the number outside of them.

In [1064]:
# Calculate the number in the ring of fire
rof = df[df.ring_of_fire == 1]
print ("Number of earthquakes in the Ring of Fire:", len(rof))

nrof = df[df.ring_of_fire == 0]
print ("Number of earthquakes outside the Ring of Fire:", len(nrof))


Number of earthquakes in the Ring of Fire: 7184
Number of earthquakes outside the Ring of Fire: 2148


## Chapter 2, Exercise 6

Find the tsunami count along the Ring of Fire.

In [1065]:
tsu_rof = rof[rof.tsunami == 1]

print ("Number of earthquakes in the Ring of Fire that caused a tsunami:", len(tsu_rof))

Number of earthquakes in the Ring of Fire that caused a tsunami: 45
